toolchain/sem_ir/function.h

```diff
@@ -60,13 +60,15 @@ struct FunctionFields {
   // implicit_param_patterns_id from EntityWithParamsBase.
   InstId self_param_id = InstId::None;
 
-  // The following member is set on the first call to the function, or at the
-  // point where the function is defined.
+  union SpecialFunctionKindInfo {
+    // If special_function_kind is Builtin, the corresponding builtin kind.
+    // This member is set at the end of a builtin function definition.
+    BuiltinFunctionKind builtin_function_kind;
 
-  // The following members are set at the end of a builtin function definition.
-
-  // If this is a builtin function, the corresponding builtin kind.
-  BuiltinFunctionKind builtin_function_kind = BuiltinFunctionKind::None;
+    // If special_function_kind is Thunk, the declaration that this is a thunk
+    // for. This member is set when a thunk is declared.
+    InstId thunk_decl_id = InstId::None;
+  } special_function_kind_info;
```

Comment:
Would this be worth changing to a `std::variant`, collapsing in the `enum`?

Reply:
My thinking is that this type will eventually be directly part of our serialized representation of SemIR, and mmap'd in when reading a serialized file, so it shouldn't contain types whose representations we don't control.